### PR TITLE
etags for small multipart uploads

### PR DESCRIPTION
**Pull request by algrs**
Add support for checking etags for multipart uploads with alternate part sizes. Previously, we only supported correctly checking etags for files we'd uploaded ourselves or that matched our chunk size. This was causing us to fail when downloading files that other tools had multipart uploaded. This adds at least some support for checking etags for multipart uploads with a part size other than 5gb. It's not perfect, as there appears to be no way to get a guaranteed result without exhaustively checking thousands of possible md5 hashes, which would be... slow. Also added support for explicitly uploading files as multipart with a custom chunk size, the better to make a test with.

### CODE REVIEW

File: baiji/s3.py [hunk 1]
```diff
@@ -880,27 +894,56 @@ def _get_etag(self, netloc, remote_path):
             raise KeyNotFound("s3://%s/%s not found on s3" % (netloc, remote_path))
         return k.etag.strip("\"") # because s3 seriously gives the md5sum back wrapped in an extra set of double quotes...
 
+    def _build_etag(self, local_path, n_parts, part_size):
```
Comment (dlsmith) [inline baiji/s3.py]: Can you add a high-level description of what you're doing here, with any relevant links you have on hand? For someone lacking context, the algorithm will seem a bit mysterious.

File: baiji/s3.py [hunk 2]
```diff
@@ -880,27 +894,56 @@ def _get_etag(self, netloc, remote_path):
...
+        if "-" not in other_etag or k.scheme == 's3':
+            return self.etag(key_or_file) == other_etag
+        else: # This is the case where the key was uploaded multipart and has a `md5-n_parts` type etag
+            n_parts = int(other_etag.split("-")[1])
+            file_size = os.path.getsize(k.path)
+            # There are a number of possible part sizes that could produce any given
+            # number of parts. The most likely and only ones we've seen so far are
+            # these, but we might someday need to try others, which might require
+            # exhaustively searching the possibilities....
+            possilbe_part_sizes = [
```
Comment (dlsmith) [inline baiji/s3.py]: s/possilbe/possible

File: baiji/s3.py [hunk 3]
```diff
@@ -880,27 +894,56 @@ def _get_etag(self, netloc, remote_path):
...
+                S3_MAX_UPLOAD_SIZE, # what we do
+                file_size/n_parts, # seen this from third party uploaders
```
Comment (dlsmith) [inline baiji/s3.py]: `ceil` and such, as we talked about offline

File: baiji/s3.py (hunk 1 above)
Comment (dlsmith) [inline baiji/s3.py]: Ah, I see this was a copy and paste from below. Still, I think it would help to have a little high-level context.

**Comment by dlsmith** [comment]
Thanks Alex 👍

**Comment by dlsmith** [comment]
Are we having the same problem with the mLab backups? Skimming the logs it looks like a similar persistent transient failure pattern.

File: baiji/s3.py (hunk 1 above)
Comment (algrs) [inline baiji/s3.py]: Will do

File: baiji/s3.py (hunk 2 above)
Comment (algrs) [inline baiji/s3.py]: thanks!

File: baiji/s3.py (hunk 3 above)
Comment (algrs) [inline baiji/s3.py]: done

**Pull request merged**